LSPS4/README.md

```diff
@@ -0,0 +1,1889 @@
+> time-bound any extra information the LSP has to store for
+> clients that request SCIDs (which are shared among all clients
+> of the same LSP) and then never actually use them.
+> The expiry time gives the client an opportunity to have a
+> channel opened (either by opening one itself, or otherwise
+> purchasing a channel from the LSP via this LSPS or via
+> [LSPS1][] or [LSPS2][]) during this expiry time period.
+
+[LSPS1]: ../LSPS1/README.md
+[LSPS2]: ../LSPS2/README.md
+
+`min_final_cltv_expiry_delta` is the final CLTV delta for the
+recipient, i.e. the `c` field in BOLT11 invoices.
+It is a positive non-zero JSON integral number.
+The LSP MUST persist this together with the generated SCID if
+this call succeeds.
+
+The client may call the `lsps4.reserve_scid` any number of
+times.
+On the second and subsequent times, if the LSP still has
+remembered a reserved SCID for this client, it MUST return the
+same SCID and its stored details.
+
+On success, the result looks like:
+
+```JSON
+{
+  "lsps4_scid": "8359321x85942x65443",
+  "expiry_time": "2023-04-19T11:02:11.841Z",
+  "min_final_cltv_expiry_delta": 24,
+  "ln_base_fee_msat": "1000",
+  "ln_prop_fee": 1000,
+  "cltv_expiry_delta": 144
+}
+```
+
+`lsps4_scid` is the SCID reserved [<LSPS0.scid>][].
+The LSP SHOULD set the highest bit of the block index (the first 24-bit
+number in the `BBBxTTTxOOO` format), and select the rest of the bits
+of the block, transaction, and output indexes by random;
+the LSP MAY use any other method of generating this SCID, and
+MAY use an SCID that could be interpreted as an actual channel, provided
+the LSP can uniquely identify the client.
+The LSP MUST ensure that the SCID is unique at the LSP, that is, no
+other peer (client or non-client) of the LSP has the same SCID allocated
+to it, whether or not the SCID could refer to an actual possible channel
+outpoint.
+
+> **Rationale** Setting the highest bit of the block index, and
+> selecting the rest by random, reduces the probability of the same
+> randomly-selected SCID, giving 63 bits of entropy.
+> The blockheight 8,388,608 is unlikely to be reached for about ~120
+> years from the time of this writing (2023).
+>
+> Regardless, the only true requirement is that the SCID refers to
+> only one possible peer, and the LSP is free to use any policy that
+> fits.
+> The LSP needs some database or lookup table for SCIDs in order to
+> implement LSPS4 anyway, and could use a simple incrementing 64-bit
+> counter that it reinterprets as an SCID with the block index as the
+> highest 24 bits (and checking it against "real" SCIDs it has with
+> peers); every actual block mined would then grant a space of
+> 1,099,511,627,776 (= 2^40) possible SCIDs, with only a tiny few of
+> them likely to become real SCIDs with peers of the LSP node, even
+> if the counter is started from 0.
+
+`expiry_time` is the expiry time for the reserved SCID
+[<LSPS0.datetime>][].
```

Comment:
We do still need an expiry, for example if the client is trying to DoS and waste the storage of the LSP by reserving and then disconnecting forever.  The LSP and the client need to agree on how long that period would be before the LSP can safely forget the client if the client never comes back again.

Reply:
Why do we need to expire the permanent SCID? Can't we just use the same SCID for a node pubkey all the time?
In this case, I don't see how the client can DDoS the LSP as there is the same SCID returned every time.

Reply:
>In this case, I don't see how the client can DDoS the LSP as there is the same SCID returned every time.

A client can always generate a new node pubkey, you can literally treat the private key as a large 256-bit number and increment it each time, no need to even use a CSPRNG if you are just using throwaway keypairs for the purpose of DoS.  With e.g. Tor or a botnet the IP address is not the same too.

Reply:
Not sure I can follow. Let's discuss it on the call today.

Reply:
Not resolved in previous call.

@SeverinAlexB proposes that an actual LSP implementation can use a 63-bit hash of the client node ID, so that the permanent SCID does not need to be remembered.

However, some software such as LDK has its own way of doing SCID interception.  LDK gets "first dibs" at the data so LSPs built on top of LDK cannot change the way that LDK creates interception SCIDs.

In addition, something like a hash is a one-way function.  So if the LSP receives an interception SCID, it has to iterate through all connected peers and hash each of their node IDs to get the SCID, then find the correct SCID.  The LSP can cache this.  However, how about disconnected peers?  Ideally if the LSP receives an SCID for a peer that is disconnected, it would wake it up (e.g. via LSPS5).  So LSP needs to remember the SCID->client mapping, and I think we need to nail down how long the mapping is retained by the LSP after the peer disconnects without making a channel.